File: pham/ws.go
package pham

import (
	"io"
	"log"
	"net/http"

	"golang.org/x/net/websocket"
)

// webSocketConnection structure
type webSocketConnection struct {
	ws *websocket.Conn
}

// Send implemented Connection interface
func (wc *webSocketConnection) Send(data []byte) (err error) {
	_, err = wc.ws.Write(data)
	return
}

// WSHandler is websocket end point
func WSHandler(w http.ResponseWriter, r *http.Request) {
	s := websocket.Server{Handler: websocket.Handler(
		func(ws *websocket.Conn) {
			connection := &webSocketConnection{ws: ws}

			// add connection
			manager.AddConnection(connection)
			defer func() {
				// delete connection
				manager.DelConnection(connection)
			}()

			for {
				// receive message
				message := new(Message)
				err := websocket.JSON.Receive(ws, message)
				if err != nil {
					// close event
					if err == io.EOF {
						return
					}

					log.Println(err)
				}
				log.Printf("client: %#v\n", message)
			}
		}),
	}
	s.ServeHTTP(w, r)
}

File: static/ws.js
var url = "ws://localhost:3000/subscribe";

var ws = new WebSocket(url);
ws.addEventListener("open", function (e) {
  console.log("ws:open:", e);
});
ws.addEventListener("close", function (e) {
  console.log("ws:close:", e);
});
ws.addEventListener("message", function (e) {
  console.log("ws:message:", e.data);
});
ws.addEventListener("error", function (e) {
  console.error("ws:error", e);
});

function send(data) {
  var json_str = JSON.stringify({
    channel: "default",
    ttl: 10,
    data: data
  });
  ws.send(json_str);
}
